### ISSUE TITLE
make check fails

### ISSUE THREAD

**Issue by sbguthery** [opencog/link-grammar#989]
Clean install of 5.6.2 on Ubuntu 18.04.2 and Python 2.7.15+.

=================================================================
   link-grammar 5.6.2: bindings/python-examples/test-suite.log
=================================================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: tests
===========

Running by: /usr/bin/python
Running ./tests.py in: /home/scott/LinkGrammar/bindings/python-examples
PYTHONPATH=./../python:../python:../python/.libs
srcdir=.
LINK_GRAMMAR_DATA=./../../data
Compiled with: gcc __VERSION__="7.4.0"
OS: linux-gnu __unix__
Standards: __STDC_VERSION__=201112L
Configuration (source code):
    CPPFLAGS=
    CFLAGS=-D_DEFAULT_SOURCE -std=c11 -D_BSD_SOURCE -D_SVID_SOURCE -D_GNU_SOURCE 
 -D_ISOC11_SOURCE -g -O2 -O3
Configuration (features):
    DICTIONARY_DIR=/usr/local/share/link-grammar
    -DPACKAGE_NAME="link-grammar" -DPACKAGE_TARNAME="link-grammar" -DPACKAGE_VERSION="5.6.2" 
 -DPACKAGE_STRING="link-grammar 
5.6.2" -DPACKAGE_BUGREPORT="https://github.com/opencog/link-grammar" -DPACKAGE_URL="https://www.abisource.com/projects/link-grammar" 
 -DPACKAGE="link-grammar" -DVERSION="5.6.2" -DYYTEXT_POINTER=1 -DSTDC_HEADERS=1 
 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 
 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1  
-DHAVE_UNISTD_H=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=".libs/" -DHAVE_STRNDUP=1 -DHAVE_STRTOK_R=1 
 -DHAVE_ALIGNED_ALLOC=1 -DHAVE_POSIX_MEMALIGN=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 
 -DHAVE_DECL_STRERROR_R=1 -DHAVE_STRERROR_R=1 -DHAVE_FORK=1 -DHAVE_VFORK=1 -DHAVE_WORKING_VFORK=1 
 -DHAVE_WORKING_FORK=1 -DHAVE_PRCTL=1 -D__STDC_FORMAT_MACROS=1 -D__STDC_LIMIT_MACROS=1 
 -DHAVE_LOCALE_T_IN_LOCALE_H=1 -DHAVE_STDATOMIC_H=1 -DTLS=__thread -DHAVE_MKLIT=1 
 -DUSE_SAT_SOLVER=1 -DUSE_WORDGRAPH_DISPLAY=1 -DHAVE_HUNSPELL=1 -DHUNSPELL_DICT_DIR="/usr/share/myspell/dicts" 
 -DHAVE_REGEX_H=1 -DHAVE_REGEXEC=1 -DHAVE_MAYBE_UNINITIALIZED=1
Using <module 'linkgrammar' from 
'/home/scott/LinkGrammar/bindings/python/linkgrammar.py'>
Using <module 'clinkgrammar' from 
'/home/scott/LinkGrammar/bindings/python/clinkgrammar.py'>
Using <module '_clinkgrammar' from 
'/home/scott/LinkGrammar/bindings/python/.libs/_clinkgrammar.so'>
Using <module 'lg_testutils' from 
'/home/scott/LinkGrammar/bindings/python-examples/lg_testutils.py'>
.............................................................link-grammar: 
Error: Could not open dictionary "demo-sql/4.0.dict"
E..........Debug: Dictionary "he/4.0.dict": Locale "he_IL.UTF-8" unknown
link-grammar: Info: Dictionary 'he/4.0.dict': No locale definition - 
"en_US.UTF-8" will be used.
.Debug: Dictionary "ru/4.0.dict": Locale "ru_RU.UTF-8" unknown
link-grammar: Info: Dictionary 'ru/4.0.dict': No locale definition - 
"en_US.UTF-8" will be used.
.ssDebug: Dictionary "de/4.0.dict": Locale "de_DE.UTF-8" unknown
link-grammar: Info: Dictionary 'de/4.0.dict': No locale definition - 
"en_US.UTF-8" will be used.
............Debug: Dictionary "lt/4.0.dict": Locale "lt_LT.UTF-8" unknown
link-grammar: Info: Dictionary 'lt/4.0.dict': No locale definition - 
"en_US.UTF-8" will be used.
.Debug: Dictionary "ru/4.0.dict": Locale "ru_RU.UTF-8" unknown
link-grammar: Info: Dictionary 'ru/4.0.dict': No locale definition - 
"en_US.UTF-8" will be used.
....
======================================================================
ERROR: setUpClass (__main__.GSQLDictTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./tests.py", line 802, in setUpClass
    cls.d, cls.po = Dictionary(lang='demo-sql'), ParseOptions()
  File "/home/scott/LinkGrammar/bindings/python/linkgrammar.py", line 375, 
in __init__
    raise LG_DictionaryError('Error: Failed to open dictionary 
{!r}'.format(lang))
LG_DictionaryError: Error: Failed to open dictionary 'demo-sql'

----------------------------------------------------------------------
Ran 90 tests in 3.689s

FAILED (errors=1, skipped=2)

**Comment by ampli** [comment]
> Error: Could not open dictionary "demo-sql/4.0.dict"

For the SQL dictionary support you need to install the SQLite development package (different names on different systems - e.g. `libsqlite3-dev` or `sqlite-devel`). Then configure and compile again.
But if you don't need it you can ignore this test failure since it was the only one that failed.

BTW, the "locale ... unknown" messages just mean you don't have these locales installed and this doesn't cause a test failure - you can just ignore that.

**Comment by ampli** [comment]
I will add something like `(may need to install the "SQLite" development package).` to this error.

**Comment by linas** [comment]
It's not enough to install it; LG needed to be compiled with the `sqlite3-dev` - the more reasonable thing to do would be to not run the unit test if the sqlite bindings were not compiled in.

**Comment by ampli** [comment]
So I propose the following change in order to be more informative and less confusing:

In `tests.py`, before performing the DB dict test, test whether the LG library is compiled with SQLlite. If not, then give an error message like `LG library is not compiled with SQLlite support - check the output of "configure"`) and skip the DB dict test.

In addition (or instead) I guess that it is also possible to add a similar error message in the code itself.

Last thing, when inspecting the code just now, I found 2 apparent problems with opening a DB dict:
1. `db_open()` in `read-sql.c`:
``` c
        /* Get the file size, in bytes. */
        /* SQLite has a habit of leaving zero-length DB's lying around */
        fd = fileno(fh);
        fstat(fd, &buf);
        fclose(fh);
        if (0 == buf.st_size)
                return NULL;
```
The intention of returning NULL is that the open will fail. But only a generic error of "could not open" will be generated, while it is possible to add a more specific message.

2. In `dictionary_create_from_db()` in `read-sql.c`:
``` c
        /* Set up the database */
        dict->db_handle = object_open(dict->name, db_open, NULL);

        dict->lookup_list = db_lookup_list;
```
Missing 'goto failure` if `dict->db_handle` is `NULL`.

I will send PRs to fix all of that.

**Comment by ampli** [comment]
Also note that on Windows the tests will still always fail because I didn't care to add SQLite build support (and even if I add, it is painful to install it and it is not usually really needed there).
So I also propose to entirely skip the DB dict test on Windows (with an appropriate test skipping message).

**Comment by linas** [comment]
Closing; I think this was fixed for version 5.7.0

**Comment by ampli** [comment]
Still in my TODO list:
- tests.py: Skip the DB dict test on Windows (with an appropriate test skipping message).
- tests.py: When not on Windows and the SQLite test fails because the library was not built with SQLite, notify that.
- Fix the library to give more details when "Unable to open dictionary" due to no SQLite support, e.g. "SQLite dict not supported".
- Fix the library to give an error message when the SQlite DB is invalid, as `sqlite3_open()` just **successfully opens any kind of garbage file**, including permission problems, and behaves, for reading, as if the DB is empty.
The current error check is a NOP (`sqlite3_open()`  always returns `SQLITE_O`).
See: https://stackoverflow.com/questions/3888529/how-to-tell-if-sqlite-database-file-is-valid-or-not
